### PR TITLE
friends-app by LGeorge007

### PR DESCRIPTION
**Pull request by LGeorge007**
### friends-app

[Demo](https://lgeorge007.github.io/friends-app/)
[Code base](https://github.com/LGeorge007/friends-app)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/LGeorge007/friends-app/script.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    controlToDefault();
+    processCard(state);
+};
+
+const processForm = function({currentTarget}){
+    state.sortByName = "none";
+    state.sortByAge = "none";
+    let selectedSex;
+    let elements = currentTarget.elements;
+    let selectedChoice = Array.from(elements.sortBy).filter(({checked}) => checked);
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:67]: Can be replaced with query selector

File: submissions/LGeorge007/friends-app/script.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    switch(selectedChoice[0].id) {
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:68]: Prefer property `value` with 'ascending`, etc. values

File: submissions/LGeorge007/friends-app/script.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+const processCard = function (state){
+    if (state.filterBySex === "all"){
+        state.gotResult = result;
+    };
+    if (state.filterBySex === "male"){
+        state.gotResult = result.filter( unit => unit.gender == "male");
+    };
+    if (state.filterBySex === "female"){
+        state.gotResult = result.filter( unit => unit.gender == "female");
+    };
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:96]: `state.gotResult = result.filter( unit => unit.gender == state.filterBySex);`

File: submissions/LGeorge007/friends-app/script.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    if (state.filterByName !== "") {
+        state.gotResult = state.gotResult.filter( elem => elem.name.first.toLowerCase().startsWith(state.filterByName) );
+    }
+
+    if (state.sortByAge === "ascending"){
+        state.gotResult.sort((a, b) => sortByAge(a, b));
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:103]:
```suggestion
        state.gotResult.sort(sortByAge);
```

File: submissions/LGeorge007/friends-app/script.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    }
+};
+let result = [];
+
+const sortByAge = function(a, b) {
+    return a.dob.age == b.dob.age
+                ? 0
+                : a.dob.age <= b.dob.age
+                        ? -1
+                        : 1;
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:29]: Don't use nested ternaries. They hard to read

File: submissions/LGeorge007/friends-app/script.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    filterByName: "",
+    sortByAge: "",
+    sortByName: "",
+    toDefault(arr) {
+        this.filterBySex = "all";
+        this.filterByName = "",
+        this.sortByAge = "ascending";
+        this.sortByName = "none";
+        this.gotResult = arr;
+    }
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:20]: Usually state should not have any functions

File: submissions/LGeorge007/friends-app/script.js [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+};
+let result = [];
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:22]: Why not in state? 🤔

File: submissions/LGeorge007/friends-app/script.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+const slideMenu = function() {
+    navPanel.classList.toggle("sidebar-toggle");
+    content.classList.toggle("content-toggle");
+}
+
+const controlToDefault = function() {
+    form.elements[1].checked = "true";
+    form.elements[9].checked = "true";
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:47]: Direct access to DOM

**Review by zonzujiro — CHANGES_REQUESTED** [review]
Leaved couple comments.

Also, I see common mistakes. Please check this list:

- Single responsibility - your functions should do only one job. As an example function in which you fetch users should only fetch them and not render, transform or process them in other ways. All this things should be done in another place, outside your function. Same applied to sort function, which usually do all types of sorting 😉
- You should handle fetch response status - https://www.tjvantoll.com/2015/09/13/fetch-and-errors/
- Don’t use names as arr, array, data or etc. They to generic. Don’t use things like sortingByABC and sortingByZYX - it’s unclear and just weird. `user1` and `user2` - don't use numbers in names. Just don't :) friendsArr also bad idea - you show data type in variable name and in JS it’s redundant
- Prefer template strings to “createElement” API
- Your template strings which contains should have intendations as in HTML
- When you need to decide which sort condition were selected, prefer value of input to “check class name of pressed element”, or “compare pressed element to saved element” etc. You can have value “female”, get value and then just use it
- Don’t overuse global variables, keep data which you are using closest as possible to place where you using it. As an example, request url can be inside function which sends request
- Redundant comments - don’t use stuff ‘———————————‘ or “here I will sort my array”. First one is ugly, second - redundant because reviewer will understand that you do sort, from your code. Only situation in which you should use comments, is when you need to describe why something is going on and this is 1% of all situations :)
- Please, doublecheck english wordings. Use code spellchecker, etc.

And remember, the best way to speed up review, is to fix errors prior review. 90% of errors are shared - go thru closed PRs with Friends App and check discussions there. You will find a places with same mistakes

File: submissions/LGeorge007/friends-app/script.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                        : 1;
+}
+
+const sortByName = function(a, b) {
+    return a.name.first.toLowerCase() == b.name.first.toLowerCase()
+                ? 0
+                : a.name.first.toLowerCase() >= b.name.first.toLowerCase()
+                        ? 1
+                        : -1;
+}
```
Comment (mustbefail) [inline submissions/LGeorge007/friends-app/script.js:38]: You can use .localeCompare(), the function returns exactly 0 1 -1. And you don't need .toLoweCase()
https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare

File: submissions/LGeorge007/friends-app/script.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    state.sortedUsers = this.unsortedUsers;
+}
+
+const sortByAge = function (a, b) {
+    return a.dob.age - b.dob.age;
+}
+
+const sortByName = function (a, b) {
+    return a.name.first.localeCompare(b.name.first);
+}
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:30]: They not sorting, they comparing ;)

File: submissions/LGeorge007/friends-app/script.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+const slideMenu = function () {
+    navPanel.classList.toggle("sidebar-toggle");
+    content.classList.toggle("content-toggle");
+}
+
+const reset = function () {
```
Comment (zonzujiro) [inline submissions/LGeorge007/friends-app/script.js:37]: Reset what? :)

**Review by zonzujiro — APPROVED** [review]
LGTM

**Pull request merged**